File: lab_1_C#/ControllerThread.cs
using System;
using System.Threading;

class ControllerThread
{
    private readonly SumThread[] threads;
    private readonly int[] delays;
    private Thread controller;

    public ControllerThread(SumThread[] threads, int[] delays)
    {
        this.threads = threads;
        this.delays = delays;
    }

    public void Start()
    {
        controller = new Thread(Run);
        controller.Start();
    }

    public void Join()
    {
        controller.Join();
    }

    private void Run()
    {
        long startTime = DateTimeOffset.Now.ToUnixTimeMilliseconds();
        bool[] stopped = new bool[threads.Length];

        while (true)
        {
            long currentTime = DateTimeOffset.Now.ToUnixTimeMilliseconds();
            bool allStopped = true;

            for (int i = 0; i < threads.Length; i++)
            {
                if (!stopped[i])
                {
                    allStopped = false;

                    if (currentTime - startTime >= delays[i])
                    {
                        threads[i].StopRunning();
                        stopped[i] = true;
                    }
                }
            }

            if (allStopped)
            {
                break;
            }
        }

        foreach (var t in threads)
        {
            t.Join();
        }
    }
}

File: lab_1/src/ControllerThread.java
public class ControllerThread extends Thread {
    private final SumThread[] threads;
    private final int[] delays;

    public ControllerThread(SumThread[] threads, int[] delays) {
        this.threads = threads;
        this.delays = delays;
    }

    @Override
    public void run() {
        Thread[] stopperThreads = new Thread[threads.length];

        for (int i = 0; i < threads.length; i++) {
            final int index = i;

            stopperThreads[i] = new Thread(() -> {
                try {
                    Thread.sleep(delays[index]);
                } catch (InterruptedException e) {
                    System.err.println("Стопер потоку #" + (index + 1) + " перерваний: " + e.getMessage());
                }
                threads[index].stopRunning();
            });

            stopperThreads[i].start();
        }

        for (SumThread thread : threads) {
            try {
                thread.join();
            } catch (InterruptedException e) {
                System.err.println("Не вдалося дочекатися завершення потоку");
            }
        }

        System.out.println("Усі потоки завершено.");
    }
}
